add <param>_0 and <param>_fit machinery back

mirca opened the pull request:
Fixes #464 and #478. 

This works fine for ``BasicPSFPhotometry``. However, we should be careful for ``IterativelySubstractedPSFPhotometry``, for instance, in case where ``positions`` is not ``None`` and ``niters > 1``,  there is no way to get initial guesses for additional parameters in the next iterations. For now, the code uses the default value of the parameter in the PSF model.

cc @eteq 

@Gabriel-p do you have time to review this? :)

Gabriel-p commented:
I'd be glad to if you can help me with some more details.

If I understand correctly, this PR fixes the #464 issue, which is related to my #440 issue (which is also related to PR #465?) So this should help me estimate an average FWHM by fitting the `sigma` in `BasicPSFPhotometry` instead of fixing it? And it will not work with `DAOPhotPSFPhotometry`?

mirca commented:
@Gabriel-p In this case, it will work with ``DAOPhotPSFPhotometry``, and the initial guess to fit ``sigma`` will be ``psf_model.sigma.value``.

eteq commented:
Re: `IterativelySubstractedPSFPhotometry`: am I understanding right that the root of the problems is that `<param>_0` has to get re-set at the start of each new iteration of the loop? What about something like having it copy `<param>_0` to a new `<param>_1` in `IterativelySubstractedPSFPhotometry`, which is used as the start of all new iterations?  Then at the end of the loop, you can set `<param>_0` back in from `<param>_1`, and delete `<param>_1`?

eteq commented:
Or even have it create new columns for every iteration, like `<param>_1`, `<param>_2`, `<param>_3`, etc.  But I'd worry that if there're a lot of iteration that the table would quickly grow to be huge...?

mirca commented:
@eteq Let's say the user wants to fit `(x_0, y_0, flux, sigma)` with `IterativelySubtractedPSFPhotometry` and he/she uses `positions` (``do_photometry`` method) to provide initial guesses. But then, in the next iterations, ``IterativelySubtractedPSFPhotometry`` uses ``self.finder`` and ``aperture_photometry`` to set up initial guesses for `(x_0, y_0)` and ``flux`` of fainter sources, respectively. However, there's no direct way to set up initial guesses for ``sigma`` (or any other parameter) in the next iterations. The solution right now is just to take the default value in the `psf_model`, i.e., `psf_model.sigma.value`, as the initial guess. 
For the case of ``sigma``, this is relatively good assuming that true value of ``sigma`` does not deviate much across the image.

But to provide a complete solution to this, I think maybe we should allow the user to pass callables to ``IterativelySubtractedPSFPhotometry`` which would be used to compute initial guesses for any additional parameters for iterations `n > 1`.

mirca commented:
For instance, let's say we would like to fit ``sigma`` as well then we would do
```
sigma_0 = 3
gaussian_prf = IntegratedGaussianPRF(sigma=sigma_0)
gaussian_prf.sigma.fixed = False

phot = IterativelySubtractedPSFPhotometry(..., psf_model=gaussian_prf, niters=2)
results = phot(image=my_image)
```
where ``sigma_0`` is the initial guess for ``sigma``, and ``results`` contains a column with ``sigma_0`` and another one with ``sigma_fit``. There is also a column called ``iter_number``, representing the number of the iteration of that estimate.

The other way to pass initial guesses is through ``positions``, i.e.,
```
results = phot(image=my_image, positions=my_positions)
```
the initial guesses for ``sigma`` should be in ``positions['sigma_0']``. For subsequent iterations, i.e. fainter sources which will be detected by the find algorithm, the value in ``gaussian_prf.sigma.value`` will be used as the initial guess.

Is that intended?

I added an example to the narrative docs.

cc @eteq, @Gabriel-p, @bsipocz, @hamogu

mirca commented:
@bsipocz Should I add a changelog entry under 0.4 or 0.3.1? I think it would be good if this can go in 0.3.1 because it fixes #478 :)

bsipocz commented:
@mirca - this is a tricky one, as I would consider this PR as a new feature that also contains a commit that fixes a bug. 
I would suggest that this should go in 0.4 (but of course @larrybradley or @eteq can veto it), and if you think the bug is serious enough to get a fix in 0.3.1 then you may want to open a separate PR that cherry picks/removes those commits with the fix from here.

mirca commented:
@bsipocz Well, I would say this PR does not bring much of a new feature, it just improves the way the output table is created. Yeah, the bug is pretty serious. So, I would prefer all this in 0.3.1, but of course I can make a PR only with the bug fix :)

And, yeah, FYI this is ready for review (although I know you all have been very busy with the release of astropy 1.3) =D

mirca commented:
pinging @eteq on this :)

eteq commented on docs/photutils/psf.rst:
```diff
@@ -481,6 +481,159 @@ Consider the previous example after the line
+-----------------------------
+
+The PSF photometry classes can also be used to fit additional parameters
+simultaneously, for instance, the PSF width and the sky background.
```
This particular use case is at first glance a bit odd.  One of the main points of PSF photometry is to use the *shape* information to get a better constraint on the flux and centering. So it would be odd to want to independently measure each sigma on its own.  The "sky background" use case is much more practical, but that's a bit more complicated to implement because there's not a natural model to use.

So maybe something like this:  
```
... to fit more model parameters than just the flux and centering.  
While a more realistic use case might be fitting sky backgrounds, 
or shape parameters of galaxies, here we use the ``sigma`` 
parameter in `IntegratedGaussianPRF` as the simplest possible
example of this feature.
```

Alternatively, you could adapt the example to use the astropy `Gaussian2D`, and have x_stddev/y_stddev/theta *all* be free parameters.  That would then sort of be like galaxy-fitting photometry. But that might get a bit overly-complex, so whichever you think is going to be clearer is fine, @mirca

mirca replied:
First option done.

eteq commented on docs/photutils/psf.rst:
```diff
@@ -481,6 +481,159 @@ Consider the previous example after the line
+
+    >>> gaussian_prf.sigma.value = 2.05
+
+Now, let's create an artificial image which has a brighter star and one
```
"artificial" -> "simulated"?

mirca replied:
Thanks!

eteq commented on docs/photutils/psf.rst:
```diff
@@ -481,6 +481,159 @@ Consider the previous example after the line
+Now, let's create an artificial image which has a brighter star and one
+overlapping companion at a distance close enough so that the detection
+algorithms won't be able to identify it, and hence we should use
+``IterativelySubtractedPSFPhotometry`` to reduce the fainter star as well.
```
"reduce" -> "find" or "detect" or "measure"?

(Usually "reducing" applies to a full dataset rather than a particular object *in* a dataset.)

mirca replied:
Thanks!

eteq commented on docs/photutils/psf.rst:
```diff
@@ -481,6 +481,159 @@ Consider the previous example after the line
+
+    import matplotlib.pyplot as plt
+    from photutils.datasets import make_random_gaussians, make_noise_image
+    from photutils.datasets import make_gaussian_sources
```
Note that the names for the `datasets` functions might be changing in #474 , so there may need to be some fixing depending on the order in which these PR's get merged.

mirca replied:
@eteq Sure, thanks for letting me know!

eteq commented on docs/photutils/psf.rst:
```diff
@@ -481,6 +481,159 @@ Consider the previous example after the line
+        sigma_0   sigma_fit       x_fit         y_fit        flux_fit
+        ------- ------------- ------------- ------------- -------------
+           2.05 2.03898434182 17.9155310586 17.0377322599 10478.1845716
+           2.05 1.77365765325 12.2809666868 19.2416219146 772.882624378
```
I don't quite understand this.  It looks like the stars were created with the *same* size, yet the fit gives *different* sizes.  That would seem to be an *incorrect* fit.  

Maybe it would be better to have the second star be a bit further away, but be a bit closer in brightness and a different gaussian size?  That way you can actually end this example with "look, it correctly fit the different sizes"

eteq replied:
(or this is irrelevant of course if you choose to do the Gaussian2D model like I said above)

mirca replied:
That's right, they are too closed together and their fluxes is also really different. So, I put them a bit further apart. Now, the results are better.

eteq commented on docs/photutils/psf.rst:
```diff
@@ -481,6 +481,159 @@ Consider the previous example after the line
+            interpolation='nearest', origin='lower')
+
+
 For more examples, also check the online notebook in the next section.
```
Because the above text has gotten a lot longer, maybe we should just drop this "For more examples" sentence?  Right now it looks like its referring to the "Fitting additional paramters" functionality, which of course isn't actually true.  And maybe change the next section to "Additional example notebooks (online)"?

mirca replied:
Done

eteq commented on photutils/psf/photometry.py:
```diff
@@ -344,6 +356,38 @@ def nstar(self, image, star_groups):
 
+    def _define_fit_param_names(self):
+        """
+        Convinience function to define mappings between the names of the
```
Minor typo: "Convinience" -> "Convenience"

mirca replied:
Thanks

eteq commented on photutils/psf/photometry.py:
```diff
@@ -274,10 +278,17 @@ def do_photometry(self, image, positions=None):
                                   sources['ycentroid'],
                                   sources['aperture_flux']])
 
+        self._pars_to_set, self._pars_to_output = self._define_fit_param_names()
```
I'm finding it a bit confusing to understand why `_pars_to_set` and `_pars_to_output` are set the way they are.  I suggest one of a few alternatives:

1. Make `_define_fit_param_names` *internally* do the setting, instead of it returning values that you then immediately set.  Then the call can just be ``self._define_fit_param_names()`` (you can leave the text in the docstrong for `_define_fit_param_names` about what these values are so that it's documented somewhere.  
2. Have `_pars_to_set` and `_pars_to_output` be only *local* variables in `do_photometry`.  I guess you'll then need to then change `nstar` to call ``self._define_fit_param_names()`` at the top, separately.
3. Have  `_pars_to_set` and `_pars_to_output` get set *when the model attribute is set*, rather than when `do_photometry` is called.  I think this is a simple as adding the snippet I've but at the bottom of this post, and removing this line. This makes more sense because then if someone, say, wants to subclass this, they don't need to remember to call `define_fit_param_names` at the top of `do_photometry` to replicate all the functionality.


```
@property
def model(self):
    return self._model
@model.setter
def model(self, val):
    self._model = val
    self._define_fit_param_names()
```

eteq replied:
Thinking about this a bit more.  It occurred to me that 3 has a failure mode that if the user changes which parameters are fixed, it renders these invalid... but there's no way for the photometry class to *know* that, because the model doesn't send any signal to the photometry class.

So that favors 1 or 2 over 3...

mirca replied:
@eteq I went for option 1 because I think it's a bit more intuitive for anyone else looking at the code. Thank you!

eteq commented on photutils/psf/photometry.py:
```diff
@@ -306,16 +317,17 @@ def nstar(self, image, star_groups):
-        star_groups = star_groups.group_by('group_id')
+        result_tab = Table()
+
+        for param_tab_name in self._pars_to_output.keys():
```
I think this means you now to change the `nstar` docstring, right?  That is, it should also say that columns like `<param>_0` are required if you have any other parameters free?

mirca replied:
Sure!

eteq commented:
oh, and I agree with @bsipocz that this is a pretty major change to call a "bug fix".  So I would advocate for most of this going in 0.4, and having the #478 be separate, if possible.

The one caveat is that this was a feature in 0.2 (sort of) that was *dropped* in 0.3 ... so nominally we are adding it back in.  But that's a tenuous argument.  I'd instead say that we think its important enough there should just be a 0.4 sooner ;)

eteq commented:
Oh, and I'm noticing another comment there about wanting a *custom* way to set parameters at different iterations, instead of a first guess.  I think this is a good idea, but it should be dealt with as a follow-on PR just so that we don't hold this one up.

As an idea for how that might work: perhaps add another attribute to the `IterativelySubtractedPSFPhotometry` class which runs right after the `finder` and can be used to adjust model parameters?  Presumably it would take the table for the current iteration,  subtracted image, and iteration number, and would be expected to add a column like `sigma_1` (if it's on iteration 1), which would then run along the lines of what I said in https://github.com/astropy/photutils/pull/471#issuecomment-264320412 ...?  Anyway, some devils in the details there, so that's why I'm suggesting a separate PR.  As a starting point I think this is fine completely punting it and just using either `<param>_0` or the default in the `psf_model`.

bsipocz commented:
@eteq - If I remember right, it wasn't released in 0.2. But I need to go back and double check it.

eteq commented:
@bsipocz oh, I think you're right.  In that case definitely 0.4

bsipocz commented:
It looks like #159 (milestoned 0.3) was the first big improvement PR on the the original (v0.1) psf, and required ``positions`` to be a table, with mandatory column names ``[xy]_0``. 

As a side note I could see benefits of letting astrobot loose on photutils to make sure we always have changelog entries, or a good excuse for not having it...

mirca commented on docs/photutils/psf.rst:
```diff
@@ -481,6 +481,159 @@ Consider the previous example after the line
     plt.colorbar(orientation='horizontal', fraction=0.046, pad=0.04)
     plt.show()
 
+Fitting additional parameters
```
Done

mirca commented:
> This does mean that positions is getting to be a pretty bad name... Really it should be initial_guesses? But perhaps it's too late to make that change if we think it breaks backwards compatibility.

@eteq ``positions`` is in fact a really bad name right now. I advocate it should be ``initial_guesses`` or ``initial_estimates``. What's the procedure to deprecate a ``keyword``? Enable, say ``initial_estimates``, and then raise a warning whenever ``positions`` is used?

bsipocz commented:
@mirca - I think the best way to do that is as it was done for  ``clobber`` vs. ``overwrite``, using the deprecated_renamed_argument decorator. However it was added in astropy 1.3, so you need to make sure that the tests are xfailed/skipped for earlier versions.

mirca commented:
@bsipocz Oh, that's cool, thanks! (:

mirca commented:
@bsipocz Travis failures looks unrelated

bsipocz commented:
@mirca - Indeed, and I've restarted the failing jobs.

@eteq - could you have a look?

eteq commented:
Haven't had a chance to do a detailed look yet... But @mirca, are you thinking to do that rename here, or as a separate PR? (i.e., are you still working on this now?)

mirca commented:
@eteq Either way (here or in another PR) works! Let me know which way do you prefer. Also, which name is best ``initial_guesses`` or ``initial_estimates`` (or any other suggestion)?

mirca commented:
@eteq @bsipocz I went ahead and renamed the keyword ``positions`` by ``init_guesses`` (let me know if you have other name in mind).

mirca commented:
> However the decorator was added in astropy 1.3, so you need to make sure that the tests are xfailed/skipped for earlier versions.

@bsipocz does that mean only ``astropy >1.3`` users would be aware of that deprecation?

bsipocz commented:
Hmm, good point. I'm not sure what's the workaround for earlier versions. One way to handle it to add a private copy of the function as it was done in astropy/astropy#3595. But I hope @eteq or @larrybradley have a better idea.

I wish we had moved ``astropy.utils`` into ``astropy_helpers``, so could always use the latest tools we have available. I wonder whether we can do it, what do you think @eteq?

eteq commented:
@bsipocz - hmm, that's an interesting idea.  The trouble is that I think *some* of the `utils` are more "astropy-y" than they are "helpers-y".  But it might be that those should be split out?  Food for thought, anyway.

For this one perhaps the solution is to just backport the astropy v1.3 to `photutils.utils` somewhere?  Along the lines of 
```
def deprecated_renamed_argument(*args, **kwargs):
    if APY_GTEQ_1.3:
        return astropy.deprecated_renamed_argument(*args, **kwargs)
    else:
        ...
```

(Will try to do a detailed review soon now that the update is in here)

mirca commented:
@eteq @bsipocz So, should I copy ``deprecated_renamed_argument`` from ``astropy`` to say ``photutils.utils.decorators`` and use that if ``astropy < 1.3``?

bsipocz commented:
@mirca @eteq - Copying over the decorator seems to be a great workaround. However in the past we did these kind of copies into ``photutils/extern``, so I would suggest to keep that tradition.

mirca commented:
@MSeifert04 what sort of black magic did you use in [#5214](https://github.com/astropy/astropy/pull/5214/) to omit the warnings in the docstrings examples? ;-)
I'm trying to copy ``deprecated_renamed_argument`` to ``photutils.extern``, but Travis complains about those omitted warnings :(

MSeifert04 commented:
@mirca I actually don't know. I suspected it was because [astropy turns deprecations into exceptions but excluding astropy-deprecations](https://github.com/astropy/astropy/blob/master/astropy/conftest.py#L14). However I have no idea why the deprecations are not shown in the doctests (maybe that's some other setting in doctest or pytest or astropy)?

Sorry, I can't help you there. Maybe just remove the doctests? 😄

mirca commented:
> I suspected it was because astropy turns deprecations into exceptions but excluding astropy-deprecations. 

Thank you @MSeifert04 =)

MSeifert04 commented:
Maybe you could also set `include_astropy_deprecations=False` in [`photultils/conftest.py`](https://github.com/astropy/photutils/blob/master/photutils/conftest.py#L9)

mirca commented:
@bsipocz Travis failures look unrelated.

bsipocz commented:
The lts failures are indeed unrelated and #503 should fix them.

mirca commented:
@eteq @bsipocz Travis looks unrelated. Is this good to go? ;-)

bsipocz commented:
@mirca - restarted travis passed 🎉. I'll leave the merging to @eteq.

mirca commented:
Thanks for restarting it, @bsipocz! =)

mirca commented:
@eteq Thanks!
